File: 1.21/smallest_divisor.rb
module SmallestDivisor
  class << self
    def calc(n)
      find_divisor(n, 2)
    end

    def find_divisor(n, divisor)
      return n if divisor ** 2 > n
      return divisor if divides?(n, divisor)
      find_divisor(n, divisor + 1)
    end

    def divides?(number, divisor)
      number % divisor == 0
    end
  end
end

File: 1.21/smallest_divisor.rs
pub fn calc(n: i64) -> i64 {
    return find_divisor(n, 2);
}

fn find_divisor(n: i64, divisor: i64) -> i64 {
    if divisor * divisor > n {
        return n;
    } else if n % divisor == 0 {
        return divisor;
    } else {
        return find_divisor(n, divisor + 1);
    }
}